### PR TITLE
feat: add Graph -> quickr lowering

### PR DESCRIPTION
**Pull request by t-kalinowski**
## Why

PJRT is great for accelerator-backed execution, but today there are two CPU-centric use cases where a `{quickr}` backend is especially compelling:

1. **CPU-only environments:** when there’s no GPU/TPU available (or we’re deploying somewhere that only has CPU), we want a fast backend that keeps execution in plain R and doesn’t rely on accelerator runtimes.
2. **AOT / caching foundation:** lowering a `Graph` to a concrete piece of R/Fortran code is a natural seam for ahead-of-time compilation and/or caching of compiled artifacts that could be easily shipped in an R package.

For pure CPU graphs, `{quickr}` should also be a strong performance baseline (and may beat PJRT-on-CPU by avoiding PJRT program/execute overhead).

In the longer term, I anticipate we'll add GPU support to quickr (likely via OpenACC, or possibly via CUDA directly), at which time `quickr` might provide a simple interface to writing custom, fast, `anvil` kernels. 

## What’s included

- New exported API: `graph_to_quickr_function()`
- Lowers a supported subset of `anvil::Graph` to a plain R function and eagerly compiles it with `quickr::quick()`
- Handles graph constants and structured (nested list) outputs by packing/unpacking values for `{quickr}`
- `{quickr}` is optional (`Suggests`); feature/tests skip gracefully when it isn’t installed
- Correctness coverage compares `{quickr}` execution vs PJRT execution

## Minimal example 

```r
library(anvil)

graph <- trace_fn(
  function(x, y) x * y + x,
  list(
    x = nv_scalar(0, dtype = "f64"),
    y = nv_scalar(0, dtype = "f64")
  )
)

f_quick <- graph_to_quickr_function(graph)
f_quick(2, 3)
#> [1] 8
```

## Integration tests

- MNIST-shaped rank-5 batch training loop: quickr-compiled `loss + grad` matches PJRT outputs, then a few SGD steps reduce the loss: 
  https://github.com/t-kalinowski/anvil/blob/quickr/tests/testthat/test-graph-to-quickr-integration.R#L1
- TFP/greta-like MAP workflow: quickr-compiled `log_prob + grad` matches PJRT, and parameter updates remain identical over multiple gradient-ascent steps:
  https://github.com/t-kalinowski/anvil/blob/quickr/tests/testthat/test-graph-to-quickr-integration.R#L70

## Notes / limitations

- Inputs must be flat (non-nested) argument lists.
- Supported primitives: `constant`, `add`, `sub`, `mul`, `divide`, `negate`, `broadcast_in_dim`, `dot_general`, `transpose`, `reshape`, `sum`.
- Current implementation supports tensors up to rank 5; `transpose` is currently rank-2 only; `sum` supports rank 0–2 reductions and full reductions for rank > 2.

### CODE REVIEW

**Review by sebffischer — COMMENTED** [review]
Awesome PR @t-kalinowski! :rocket: I didn't really dig into the quickr emitting code, because I don't really know how Fortran works. I hope I can look into this more.

I have some more comments:

1. I am wondering whether this is better suited in its own package as it seems similar in spirit to [stablehlo](https://github.com/r-xla/stablehlo). 
1. I think in combination with https://github.com/r-xla/pjrt/commit/e787de9c655b6643ce0642c66dc6761ff22d4cfd, which allows to register custom functions with PJRT, we might be able to define custom kernels in quickr and even use them within an XLA program, is this correct @dfalbel? (We would only have to be careful because XLA uses row-major layout for arrays.)
1. I think it would make sense to have (almost completely automated) unit tests for the primitives that check that they are identical between the XLA backend and the quickr backend like we do with XLA and torch [here](https://github.com/t-kalinowski/anvil/blob/aba7f0beef4f019e5c4699491e8362a9994d5a06/inst/extra-tests/test-primitives-stablehlo-torch.R#L13-L15). (I can also help with this)
1. I think currently literals are not handled, for which I introduced support for in anvil [this week](https://github.com/r-xla/anvil/commit/477ea167e6d9b5b735377cd7b1aa741bab74be07). I.e., the inputs of a `PrimitiveCall` can now also be a [`GraphLiteral`](https://github.com/r-xla/anvil/blob/6ae09adfefaa70ba4e24182657b1ace8f9124fe9/R/graph.R#L22-L27) which holds a [`LiteralTensor`](https://github.com/r-xla/anvil/blob/6ae09adfefaa70ba4e24182657b1ace8f9124fe9/R/tensor.R#L203-L230).
Consider this:

    ```r
    f <- function(x) x + 1
    graph <- trace_fn(f, list(nv_scalar(1))
    fun <- graph_to_quickr_r_function(graph)
    ```

Also a TODO for me:

* [ ] `trace_fn` should not require `AnvilTensor` inputs but also work with `AbstractTensor` inputs
* [ ] the `compile_graph_pjrt` function from the test helpers should be defined by the anvil package itself as it's generally useful

**Comment by sebffischer** [comment]
> > I think in combination with [r-xla/pjrt@e787de9](https://github.com/r-xla/pjrt/commit/e787de9c655b6643ce0642c66dc6761ff22d4cfd), which allows to register custom functions with PJRT, we might be able to define custom kernels in quickr and even use them within an XLA program, is this correct @dfalbel? (We would only have to be careful because XLA uses row-major layout for arrays, as well as type conversions (we would also need more types like f32 in quickr I think).
> 
> While that could be possible I don't it's worth it because the execution models are so different.
What do you mean by that?

> I think what could be nice is if anvil allowed users to register new specific ops, eg currently we need:
> 
> ```
> relu <- function(x) {
>    nv_max(x, 0)
> } 
> 
> function(x) {
>   relu(relu(x) * 2  +b)
> }
> ```
> 
> But in principle, a user could bring their own implementation of rely in quick that doesn't use `max` and could be potentially faster.
> 
> So users should be able to tell anvil they have a primitive for `relu()` for instance.

That would be nice indeed. Currently one would have to create a new primitive graph-building function `nvl_reul` and register the appropriate rules, which is good enough for now I think.

**Comment by sebffischer** [comment]
I also noticed that the dtype `"f32"` is simply ignored by the quickr backend:

```r
> graph <- trace_fn(\(x, y) x * y, list(x = nv_aten("f32", 1L), y = nv_aten("f32", 1L)))
> graph_to_quickr_r_function(graph=graph)
function (x1, x2) 
{
    declare(x1 = type(x1 = double(1L)), x2 = type(x2 = double(1L)))
    anvil_quickr_v1 <- x1 * x2
    anvil_quickr_out <- array(anvil_quickr_v1, dim = 1L)
    anvil_quickr_out
}
<environment: 0x116151b30>
```

This should throw an error I think.

**Comment by sebffischer** [comment]
Also just for later: I wonder whether it's better to have the AnvilPrimitives broadcast scalars by default. Currently we call into `broadcast_in_dim` even for scalars but if we define the semantics of the primitive binary ops to already be broadcasted (`nvl_add(1, c(1, 2))` would work, not only `nv_add(1, c(1, 2))`) then the quickr backend could have a specialized code path for adding a scalar to an n-dimensional array which might be more efficient than calling into the broadcasting function, but maybe the Fortran compiler is smart enough)
